src/net_processing.cpp

```diff
@@ -2433,9 +2433,8 @@ void ProcessMessage(
         UpdatePreferredDownload(pfrom, State(pfrom.GetId()));
         }
 
-        if (!pfrom.IsInboundConn() && pfrom.IsAddrRelayPeer())
+        if (pfrom.AdvertiseAddressToConn())
```

Comment:
If we're cleaning this up, we should really clean it up and move the `connman.MarkAddressGood()` out of this if statement, since it's not related to address advertising at all. Having it in the same conditional is a trap for developers to fall into.

(specifically, a developer might think "we don't need to advertise our address to FEELER connections, since we disconnect before we actually send the ADDR or GETADDR messages, so we can remove FEELER from AdvertiseAddressToConn", but doing that breaks our FEELER connection logic, because `MarkAddressGood()` needs to be called to mark the address as good)

cc @sdaftuar

Reply:
I was looking at this more and I believe we should leave the MarkAddressGood call in the if-block, though it needs an explanation— the reason (I think) is that we don’t want block-relay connections to leave a trace in the addrman that could be leaked by an attacker sniffing our addr responses.

So this just needs a comment explaining why it’s in here.

Reply:
Can you explain this a bit? I think you're saying that if we connect out to a peer that we use for block-relay-only we don't want to mark it as good (essentially move it from New to Tried and update the last successful connect time) because then another peer might be able to tell that we've connected to them?

My understanding of how we respond to `getaddr` messages is that we'll respond with (up to) 1000 records taken at random from all records, regardless of whether they're in New or Tried, so marking a peer as good won't change whether we include it in `addr` responses.

Reply:
I don't have a great understanding of the details of how addrman works or how we interact with it -- so possibly the fear was unfounded.  It used to be the case that you could trivially download a listening node's addrman though, and it also used to be the case (perhaps only long ago?) that we'd leak information about current connections based on time stamps stored in our addrman, so when I worked on #15759 I tried to [leave the addrman state unchanged](https://github.com/bitcoin/bitcoin/pull/15759#issuecomment-480398802) by making block-relay-only connections.

I'm not sure that I fully succeeded in not leaking information, or that the concern is correct to begin with given how addrman already works.  So it's possible that with some analysis we could make improvements here.  Maybe not appropriate for this pr though...

Reply:
I believe there are three ways that a peer's record is updated in addrman when we're connected to them:

1. when we receive a `version` message from them, we'll mark their address good. That moves the address from the new table to the tried table, and updates the `nLastSuccess`/`nLastTry`/`nAttempts` fields. We only do this if the peer isn't inbound and isn't block-relay-only. I don't believe this change can easily be detected by a peer sending us `getaddr` messages.
2. if a peer self-advertises to us, we'll call `AddAddressKnown()`, which could update the `nTime`, which is included in the address record that we send to a peer in response to a `getaddr` request.
3. when we disconnect from the peer, we'll update the `nTime`, as long as the peer was marked as `fCurrentlyConnected`. That happens when we receive a `verack` from a peer which is not inbound. This doesn't include feeler peers, since we disconnect from them as soon as we receive a `version` message from them. The `nTime` is only updated during disconnection so spies can't find out who we're currently connected to (https://github.com/bitcoin/bitcoin/pull/5860#issuecomment-78397140).

In summary, I think block-relay-only peers _should_ have their address marked good when we connect to them (1). If we don't, then we'll update the `nLastTry` and `nAttempts` fields when we attempt to connect to them, but not update the `nLastSuccess`.

The reason it's important in this PR is that placing `MarkAddressGood()` behind a conditional `if (AdvertiseAddressToConn())` is very confusing for anyone reading the code, and makes it easy to introduce subtle bugs.

Reply:
I agree that having `MarkAddressGood` in this conditional is surprising & that changing to `AdvertiseAddressToConn` augments the confusion. 

For this PR- I've updated the conditional to be `!pfrom.IsInboundConn() && !pfrom.IsBlockOnlyConn()` (aka removed the misleading function), and added a comment.

For future work- I've noted this down. I've taken an initial look and @jnewbery I believe what you're saying makes sense, but I'll have to spend more time with it to fully convince myself. The fact that I missed this behavior in my initial pass affirms that its easy to miss or confuse. I'd like to either (1) understand why it should be in this conditional & comment or (2) understand why it shouldn't be & change it. But I'm leaving this for another PR :)

Reply:
I think we're taking the wrong approach in the PR as it is right now. *At least* that condition should be split in 2 blocks:

I think the current code change here is not useful, but only distracts:
I) The comment `Advertise our address` is removed for no reason
II) The comment `(unrelated to advertising our address)` is very confusing. I managed to understand it only after reading this discussion
III) The comment `Moves address from New to Tried table in Addrman` is [not sufficient](https://github.com/bitcoin/bitcoin/pull/19724/files/c343ba35cfecb8aaf3aecb1259b6b30a225507f3..f26b0ddac590ac96e86fe043de744747e834ec0e#r480969871)
IV) The change of `IsAddrRelayPeer` to `IsBlockOnlyConn` [here](https://github.com/bitcoin/bitcoin/pull/19724/files/c343ba35cfecb8aaf3aecb1259b6b30a225507f3..f26b0ddac590ac96e86fe043de744747e834ec0e#r480966499) is very confusing, the two blocks should be split into 2 very separate blocks according to the following reasoning:

1. We self-announce to every peer that supports self-announcement AND it's a good idea:
1a) not useful to self-announce to feelers as we won't actually send ADDR to them (i think?)
1b) bad to self-announce to block-relay-only (maybe? wrt privacy)
2. We MarkAddressGood according to the logic we have

Reply:
w.r.t block-relay-only peers and AddrMan, I agree it's a separate issue, so i'd prefer to preserve the approach status quo here.

The current PR currently satisfies it, but I'm still unhappy with the code changes in this particular area (see the previous comment).

Reply:
@naumenkogs Would it be enough to just add some more comments to this block?  I think it's fine to leave as one block until we actually need to split it, so I'd propose something like this:

```diff
diff --git a/src/net_processing.cpp b/src/net_processing.cpp
index 7cd4f7c4cb5..4285e8ba898 100644
--- a/src/net_processing.cpp
+++ b/src/net_processing.cpp
@@ -2427,6 +2427,21 @@ void PeerLogicValidation::ProcessMessage(CNode& pfrom, const std::string& msg_ty
         }
 
         if (!pfrom.IsInboundConn() && !pfrom.IsBlockOnlyConn()) {
+            // For outbound peers, we try to relay our address (so that other
+            // nodes can try to find us more quickly, as we have no guarantee
+            // that an outbound peer is even aware of how to reach us) and do a
+            // one-time address fetch (to help populate/update our addrman). If
+            // we're starting up for the first time, our addrman may be pretty
+            // empty and no one will know who we are.
+            //
+            // We also will update the addrman to record connection success for
+            // these peers (which include OUTBOUND_FULL_RELAY peers and FEELER
+            // connections) so that addrman will have an up-to-date notion of
+            // which peers are online and available.
+            //
+            // We skip these operations for BLOCK_RELAY peers to avoid
+            // potentially leaking information about our BLOCK_RELAY
+            // connections via the addrman or address relay.
             if (fListen && !::ChainstateActive().IsInitialBlockDownload())
             {
                 CAddress addr = GetLocalAddress(&pfrom.addr, pfrom.GetLocalServices());
@@ -2446,8 +2461,8 @@ void PeerLogicValidation::ProcessMessage(CNode& pfrom, const std::string& msg_ty
             m_connman.PushMessage(&pfrom, CNetMsgMaker(nSendVersion).Make(NetMsgType::GETADDR));
             pfrom.fGetAddr = true;
 
-            // Moves address from New to Tried table in Addrman
-            // (unrelated to advertising our address)
+            // Moves address from New to Tried table in Addrman, resolves
+            // tried-table collisions, etc.
             m_connman.MarkAddressGood(pfrom.addr);
         }
```

Reply:
I'd suggest making `If we're starting up for the first time, our addrman may be pretty empty and no one will know who we are.`  less ambiguous:
`If we're starting up for the first time, it would improve our peering chances. Otherwise, our addrman may be pretty empty, and also no one will know who we are.`

But generally, I think this comment:
- makes it much more clear
- actually convinces me that it's an improvement over the existent code.

Reply:
incorporated these comments into the changes, 

I agree this should be further improved, but I'm hoping this is sufficient for this PR. @naumenkogs are you comfortable with the current state?